google/cloud/spanner_dbapi/client_side_statement_executor.py

```diff
@@ -0,0 +1,34 @@
+# Copyright 20203 Google LLC All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from google.cloud.spanner_dbapi.parsed_statement import (
+    ParsedStatement,
+    ClientSideStatementType,
+)
+
+
+class StatementExecutor(object):
+    def __init__(self, connection):
+        self.connection = connection
+
+    def execute(self, parsed_statement: ParsedStatement):
+        """Executes the client side statements by calling the relevant method
+
+        :type parsed_statement: ParsedStatement
+        :param parsed_statement: parsed_statement based on the sql query
+        """
+        if (
```

Comment:
Random Python-ism:  This looks kindof like a switch/case statement.  If you find yourself implementing a switch/case statement with a large number of cases, one way to improve performance (whether it's actually more readable is subjective) is to construct a dict() mapping the possible values to functions implementing the handler logic for each value.  Then do something like `statement_type_actions[parsed_statement.client_side_statement_type](self.connection)`.

But that's certainly overkill for a single type of statement.  Personally, I wouldn't bother unless you have at least a couple dozen statement types to check; maybe more.  For anything smaller, just chain a bunch of `elif`'s.  Just mentioning it for future reference.

The modern way to do switch/case in Python would be [Structural Pattern Matching](https://peps.python.org/pep-0636/).  But this project currently supports Python 3.7+ (released in mid 2018) and Structural Pattern Matching was I believe released in Python 3.10 (released in late 2021).  So taking a dependency on it would be a big update for us.  I believe Python has a 5-year support lifecycle, so this might make more sense to discuss in 2026 or so.

(Tangent, on that note:  I believe Python 3.7 has just recently run past its support cycle?  Google generally frowns upon supporting, testing, or otherwise running software that no longer receives security updates.  So it might be worth starting to think about what the process looks like to bump this package up to a minimum dependency of Python 3.8.  Mechanically, it involves updating `setup.py` in the top-level directory to list a new minimum Python version -- though I'll note that the whole concept of `setup.py` is now deprecated and replaced with `setup.toml`; also I imagine some customers may care about the deprecation of support for a version of Python that they might still be using.)

Reply:
Thanks, I also looked into how to use switch/case in python for current python versions and found there is no native support for it. We might need to refactor this class to have config file similar to java and use reflection so didn't choose to create a dict(). Let me know if this sounds good

Reply:
Sounds good to me.  I think the current code is fine for now, and any of the proposed changes here could make sense depending on how the code evolves over time.  Reflection is easy and common in Python (IMO even more so than in Java) so a reflection-based approach seems to me like it could also make sense if the preference is to follow the precedent that's already set in Java.

Reply:
Sure, thanks